### ISSUE TITLE
CUSOLVER_STATUS_EXECUTION_FAILED when using `torch.linalg.solve`

### ISSUE THREAD

**Issue by gleize** [pytorch/pytorch#70122]
### 🐛 Describe the bug

I noticed a potential bug with the PyTorch linear solver.
It triggers a crash when using GPU only.
It works fine on CPU, or when using an older version of PyTorch (1.10.0+cu111).

### Relevant informations for reproducing the bug

python version : 3.8.12 (installed via conda)
cudatoolkit version : 11.3.1 (installed using conda)
kornia version : 0.6.2 (installed using pip)
torch version : 1.10.0+cu113 (installed using pip)
nvidia driver version : 450.80.02
error : `CUSOLVER_STATUS_EXECUTION_FAILED`
stacktrace :
```
Traceback (most recent call last):
  File "tmp.py", line 23, in <module>
    mat = get_perspective_transform(src, dest)
  File "<home>/.local/lib/python3.8/site-packages/kornia/geometry/transform/imgwarp.py", line 363, in get_perspective_transform
    X, _ = _torch_solve_cast(b, A)
  File "<home>/.local/lib/python3.8/site-packages/kornia/utils/helpers.py", line 124, in _torch_solve_cast
    out = solve(A.to(dtype), input.to(dtype))
RuntimeError: cusolver error: CUSOLVER_STATUS_EXECUTION_FAILED, when calling `cusolverDnDgetrf( handle, m, n, dA, ldda, static_cast<double*>(dataPtr.get()), ipiv, info)`
```

### Code

```python
import torch
from kornia.geometry.transform import get_perspective_transform

device = "cuda:0"
# device = "cpu"
dtype = torch.float64

points = [[[-1.0, -1.0], [1.0, -1.0], [-1.0, 1.0], [1.0, 1.0]]]

dest = torch.tensor(
    points,
    device=device,
    dtype=dtype,
)

src = torch.tensor(
    points,
    device=device,
    dtype=dtype,
)

# crashes here, should return identity matrix
mat = get_perspective_transform(src, dest)
```

### Versions

```
Collecting environment information...
PyTorch version: 1.10.0+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8.12 (default, Oct 12 2021, 13:49:34)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-52-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: Quadro GP100
GPU 1: Quadro GP100

Nvidia driver version: 450.80.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.4
[pip3] pytorch-lightning==1.5.2
[pip3] torch==1.10.0+cu113
[pip3] torch-tb-profiler==0.3.1
[pip3] torchmetrics==0.6.0
[pip3] torchvision==0.11.1+cu113
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mypy-extensions           0.4.3                    pypi_0    pypi
[conda] numpy                     1.21.4                   pypi_0    pypi
[conda] pytorch-lightning         1.5.2                    pypi_0    pypi
[conda] torch                     1.10.0+cu113             pypi_0    pypi
[conda] torch-tb-profiler         0.3.1                    pypi_0    pypi
[conda] torchmetrics              0.6.0                    pypi_0    pypi
[conda] torchvision               0.11.1+cu113             pypi_0    pypi
```

cc @ezyang @gchanan @zou3519 @bdhirsh @ngimel @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

**Comment by Lezcano** [comment]
What's your CUDA toolkit version?

**Comment by gleize** [comment]
> What's your CUDA toolkit version?

This one : `[conda] cudatoolkit               11.3.1               h2bc3f7f_2`

**Comment by Lezcano** [comment]
Could you perhaps print what are the matrices that are being factored in
>  File "<home>/.local/lib/python3.8/site-packages/kornia/utils/helpers.py", line 124, in _torch_solve_cast
    out = solve(A.to(dtype), input.to(dtype))
?

`torch.linalg.solve` just works for matrices `A` that are invertible. I think that in this case the matrix you might be using may not be invertible.

**Comment by melgor** [comment]
I had the same issue. I found some solution for that. 
At my case the problem was CUDA version installed at VM (I was using GCP VM), also maybe the driver version.

At problematic VM I have (VM-1):
```
Mon Jan 17 08:05:29 2022
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 450.80.02    Driver Version: 450.80.02    CUDA Version: 11.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   69C    P0    20W /  70W |      0MiB / 15109MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
```

At working one (VM-2):
```
Mon Jan 17 08:06:18 2022
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.73.01    Driver Version: 460.73.01    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   68C    P0    31W /  70W |      0MiB / 15109MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
```

As we can see, we have different CUDA version and different NVIDIA driver version.

In both VM I created fresh env using:
```
conda  create -n kornia python=3.8
pip3 install torch==1.10.1+cu113 torchvision==0.11.2+cu113 torchaudio==0.10.1+cu113 -f https://download.pytorch.org/whl/cu113/torch_stable.html
pip install kornia
```

After further investigation, I detected than when I will install pytorch for CUDA 11.1 in VM-1 (problematic one), everything start to work fine.
```
pip3 install torch==1.10.1+cu111 torchvision==0.11.2+cu111 torchaudio==0.10.1+cu111 -f https://download.pytorch.org/whl/cu111/torch_stable.html
```

In summary, 2 solutions:
1. Upgrade your VM cuda to at least 11.2
2. Use torch for CUDA 11.1 instead of 11.3

> Could you perhaps print what are the matrices that are being factored in
> 
> > File "/.local/lib/python3.8/site-packages/kornia/utils/helpers.py", line 124, in _torch_solve_cast
> > out = solve(A.to(dtype), input.to(dtype))
> > ?
> 
> `torch.linalg.solve` just works for matrices `A` that are invertible. I think that in this case the matrix you might be using may not be invertible.

This is not the case for this reproduce script as when you run it as CPU, everything works smoothly.

**Comment by Lezcano** [comment]
Even then, it'd be good to know which matrix is this failing on, because it might be a matrix with 1e-8 as a singular value, and different algorithms may behave differently for such a matrix. Could you save the matrix you are using in the solver so that we can have a look at it?

**Comment by ngimel** [comment]
At this point there's not much actionable things for us, please ping when there's a repro

**Comment by gleize** [comment]
This is the actual solve operation `get_perspective_transform` is doing under the hood (sorry for the late reply).

```python
import torch

A = torch.tensor(
    [
        [
            [-1.0, -1.0, 1.0, 0.0, 0.0, 0.0, -1.0, -1.0],
            [0.0, 0.0, 0.0, -1.0, -1.0, 1.0, -1.0, -1.0],
            [1.0, -1.0, 1.0, 0.0, 0.0, 0.0, -1.0, 1.0],
            [0.0, 0.0, 0.0, 1.0, -1.0, 1.0, 1.0, -1.0],
            [-1.0, 1.0, 1.0, 0.0, 0.0, 0.0, -1.0, 1.0],
            [0.0, 0.0, 0.0, -1.0, 1.0, 1.0, 1.0, -1.0],
            [1.0, 1.0, 1.0, 0.0, 0.0, 0.0, -1.0, -1.0],
            [0.0, 0.0, 0.0, 1.0, 1.0, 1.0, -1.0, -1.0],
        ]
    ],
    device="cuda:0",
    dtype=torch.float64,
)

B = torch.tensor(
    [[[-1.0], [-1.0], [1.0], [-1.0], [-1.0], [1.0], [1.0], [1.0]]],
    device="cuda:0",
    dtype=torch.float64,
)

C = torch.linalg.inv(A)

print(C)

out = torch.linalg.solve(A, B)

print(out)
```
So as you can see, A is invertible.

One thing that is quite disturbing is that the initial code and this new one now works with the exact same environment. The only thing that changed is the NVIDIA driver version (went from `450.80.02` to `470.57.02`).

@[Lezcano](https://github.com/Lezcano)

**Comment by Lezcano** [comment]
`A` is indeed invertible in that example. I run that code in master and it works for me. It certainly looks like it's coming from some bug in some previous version of cusolver... cc @IvanYashchuk @xwang233 do you know what could it be?

Also, @gleize, do you think that perhaps the matrix generated with the previous set-up was ill-defined and this one isn't?

**Comment by IvanYashchuk** [comment]
With driver version `465.19.01`, CUDA 11.3, and the latest master branch, I don't see the failure for the above code.

There could be a bug in the combination of the CUDA Toolkit version and the driver which seems fixed in the latest versions. Unfortunately, I don't think we can do much here unless the problem starts to appear with the latest versions of CUDA, its driver, and PyTorch.

**Comment by gleize** [comment]
> `A` is indeed invertible in that example. I run that code in master and it works for me. It certainly looks like it's coming from some bug in some previous version of cusolver... cc @IvanYashchuk @xwang233 do you know what could it be?
> 
> Also, @gleize, do you think that perhaps the matrix generated with the previous set-up was ill-defined and this one isn't?

@Lezcano No. This last matrix is the exact same matrix as in the previous set-up.

**Comment by scabini** [comment]
I'm having the same issue when trying to do ```torch.pinverse()``` in an invertible matrix.
Works fine with ```Nvidia driver 470.141.03 + CUDA 11.4```, while gives the error with ```Nvidia driver 450.102.04 + CUDA 11.0```

**Comment by Lezcano** [comment]
Alas, as Ivan mentioned, there's not much that we can do. This seems to be an issue with older versions fo the NVIDIA  drivers and CUDA. To whoever is facing this issue, the recommendation would be to update these to a newer version.
